File: rust/src/red_hat_boy.rs
use web_sys::HtmlImageElement;

use crate::{red_hat_boy_states::{RedHatBoyState, Idle, Running, Sliding, Jumping, Falling, KnockedOut, RedHatBoyContext, SlidingEndState, JumpingEndState, FallingEndState}, engine::{Sheet, Cell, Renderer, Rect}};

pub enum Event {
	Run,
	Slide,
	Update,
	Jump,
	KnockOut,
	Land(i16),
}

#[derive(Copy, Clone)]
enum RedHatBoyStateMachine {
	Idle(RedHatBoyState<Idle>),
	Running(RedHatBoyState<Running>),
	Sliding(RedHatBoyState<Sliding>),
	Jumping(RedHatBoyState<Jumping>),
	Falling(RedHatBoyState<Falling>),
	KnockedOut(RedHatBoyState<KnockedOut>),
}

impl RedHatBoyStateMachine {
	fn transition(self, event: Event) -> Self {
			match (self, event) {
					(RedHatBoyStateMachine::Idle(state), Event::Run) => state.run().into(),
					(RedHatBoyStateMachine::Idle(state), Event::Update) => state.update().into(),
					(RedHatBoyStateMachine::Running(state), Event::Land(position)) => {
							state.land_on(position).into()
					}
					(RedHatBoyStateMachine::Running(state), Event::Slide) => state.slide().into(),
					(RedHatBoyStateMachine::Running(state), Event::KnockOut) => state.knock_out().into(),
					(RedHatBoyStateMachine::Running(state), Event::Update) => state.update().into(),
					(RedHatBoyStateMachine::Running(state), Event::Jump) => state.jump().into(),
					(RedHatBoyStateMachine::Sliding(state), Event::Land(position)) => {
							state.land_on(position).into()
					}
					(RedHatBoyStateMachine::Sliding(state), Event::Update) => state.update().into(),
					(RedHatBoyStateMachine::Sliding(state), Event::KnockOut) => state.knock_out().into(),
					(RedHatBoyStateMachine::Jumping(state), Event::Land(position)) => {
							state.land_on(position).into()
					}
					(RedHatBoyStateMachine::Jumping(state), Event::Update) => state.update().into(),
					(RedHatBoyStateMachine::Jumping(state), Event::KnockOut) => state.knock_out().into(),
					(RedHatBoyStateMachine::Falling(state), Event::Update) => state.update().into(),
					_ => self,
			}
	}

	fn frame_name(&self) -> &str {
			match self {
					RedHatBoyStateMachine::Idle(state) => state.frame_name(),
					RedHatBoyStateMachine::Running(state) => state.frame_name(),
					RedHatBoyStateMachine::Sliding(state) => state.frame_name(),
					RedHatBoyStateMachine::Jumping(state) => state.frame_name(),
					RedHatBoyStateMachine::Falling(state) => state.frame_name(),
					RedHatBoyStateMachine::KnockedOut(state) => state.frame_name(),
			}
	}

	fn context(&self) -> &RedHatBoyContext {
			match self {
					RedHatBoyStateMachine::Idle(state) => &state.context(),
					RedHatBoyStateMachine::Running(state) => &state.context(),
					RedHatBoyStateMachine::Sliding(state) => &state.context(),
					RedHatBoyStateMachine::Jumping(state) => &state.context(),
					RedHatBoyStateMachine::Falling(state) => &state.context(),
					RedHatBoyStateMachine::KnockedOut(state) => &state.context(),
			}
	}

	fn update(self) -> Self {
			self.transition(Event::Update)
	}
}

impl From<RedHatBoyState<Idle>> for RedHatBoyStateMachine {
	fn from(state: RedHatBoyState<Idle>) -> Self {
			RedHatBoyStateMachine::Idle(state)
	}
}

impl From<RedHatBoyState<Running>> for RedHatBoyStateMachine {
	fn from(state: RedHatBoyState<Running>) -> Self {
			RedHatBoyStateMachine::Running(state)
	}
}

impl From<RedHatBoyState<Sliding>> for RedHatBoyStateMachine {
	fn from(state: RedHatBoyState<Sliding>) -> Self {
			RedHatBoyStateMachine::Sliding(state)
	}
}

impl From<SlidingEndState> for RedHatBoyStateMachine {
	fn from(end_state: SlidingEndState) -> Self {
			match end_state {
					SlidingEndState::Complete(running_state) => running_state.into(),
					SlidingEndState::Sliding(sliding_state) => sliding_state.into(),
			}
	}
}

impl From<RedHatBoyState<Jumping>> for RedHatBoyStateMachine {
	fn from(state: RedHatBoyState<Jumping>) -> Self {
			RedHatBoyStateMachine::Jumping(state)
	}
}

impl From<JumpingEndState> for RedHatBoyStateMachine {
	fn from(end_state: JumpingEndState) -> Self {
			match end_state {
					JumpingEndState::Complete(running_state) => running_state.into(),
					JumpingEndState::Jumping(jumping_state) => jumping_state.into(),
			}
	}
}

impl From<RedHatBoyState<Falling>> for RedHatBoyStateMachine {
	fn from(state: RedHatBoyState<Falling>) -> Self {
			RedHatBoyStateMachine::Falling(state)
	}
}

impl From<FallingEndState> for RedHatBoyStateMachine {
	fn from(end_state: FallingEndState) -> Self {
			match end_state {
					FallingEndState::Complete(running_state) => running_state.into(),
					FallingEndState::Falling(sliding_state) => sliding_state.into(),
			}
	}
}

impl From<RedHatBoyState<KnockedOut>> for RedHatBoyStateMachine {
	fn from(state: RedHatBoyState<KnockedOut>) -> Self {
			RedHatBoyStateMachine::KnockedOut(state)
	}
}

pub struct RedHatBoy {
	state_machine: RedHatBoyStateMachine,
	sprite_sheet: Sheet,
	image: HtmlImageElement,
}

impl RedHatBoy {
	pub fn new(sheet: Sheet, image: HtmlImageElement) -> Self {
			RedHatBoy {
					state_machine: RedHatBoyStateMachine::Idle(RedHatBoyState::new()),
					sprite_sheet: sheet,
					image,
			}
	}

	fn frame_name(&self) -> String {
			format!(
					"{} ({}).png",
					self.state_machine.frame_name(),
					(self.state_machine.context().frame / 3) + 1
			)
	}

	fn current_sprite(&self) -> Option<&Cell> {
			self.sprite_sheet.frames.get(&self.frame_name())
	}

	pub fn draw(&self, renderer: &Renderer) {
			let sprite = self.current_sprite().expect("Cell not found");
			renderer.draw_image(
					&self.image,
					&Rect::new_from_x_y(
							sprite.frame.x.into(),
							sprite.frame.y.into(),
							sprite.frame.w.into(),
							sprite.frame.h.into(),
					),
					&self.destination_box(),
			);
	}

	fn destination_box(&self) -> Rect {
			let sprite = self.current_sprite().expect("Cell not found");
			Rect::new_from_x_y(
					(self.state_machine.context().position.x + sprite.sprite_source_size.x).into(),
					(self.state_machine.context().position.y + sprite.sprite_source_size.y).into(),
					sprite.frame.w.into(),
					sprite.frame.h.into(),
			)
	}

	pub fn bounding_box(&self) -> Rect {
			const X_OFFSET: i16 = 18;
			const Y_OFFSET: i16 = 14;
			const WIDTH_OFFSET: i16 = 28;
			let mut bounding_box = self.destination_box();
			bounding_box.set_x(bounding_box.x() + X_OFFSET);
			bounding_box.width -= WIDTH_OFFSET;
			bounding_box.set_y(bounding_box.y() + Y_OFFSET);
			bounding_box.height -= Y_OFFSET;
			bounding_box
	}

	pub fn update(&mut self) {
			self.state_machine = self.state_machine.update();
	}

	pub fn run_right(&mut self) {
			self.state_machine = self.state_machine.transition(Event::Run);
	}

	pub fn slide(&mut self) {
			self.state_machine = self.state_machine.transition(Event::Slide);
	}

	pub fn jump(&mut self) {
			self.state_machine = self.state_machine.transition(Event::Jump);
	}

	pub fn land_on(&mut self, position: i16) {
			self.state_machine = self.state_machine.transition(Event::Land(position));
	}

	pub fn knock_out(&mut self) {
			self.state_machine = self.state_machine.transition(Event::KnockOut);
	}

	pub fn pos_y(&self) -> i16 {
			self.state_machine.context().position.y
	}

	pub fn velocity_y(&self) -> i16 {
			self.state_machine.context().velocity.y
	}

	pub fn walking_speed(&self) -> i16 {
			self.state_machine.context().velocity.x
	}
}

File: js/src/red_hat_boy.ts
import { ICell, ISheet, Rect, Renderer, Sound, Audio } from "./engine";
import {
  IdleRedHatBoyState,
  JumpingRedHatBoyState,
  RedHatBoyContext,
  RedHatBoyState,
  RunningRedHatBoyState,
  SlidingRedHatBoyState,
  StateType,
} from "./red_hat_boy_states";

enum Event {
  Run,
  Slide,
  Update,
  Jump,
  KnockOut,
  Land,
}

class RedHatBoyStateMachine {
  private state: RedHatBoyState;

  constructor(state: RedHatBoyState) {
    this.state = state;
  }

  transition(event: Event, ...args: any[]): RedHatBoyStateMachine {
    if (this.state.type === StateType.Idle && event === Event.Run) {
      this.state = (<IdleRedHatBoyState>this.state).run();
    }
    if (this.state.type === StateType.Running && event === Event.Land) {
      this.state = (<RunningRedHatBoyState>this.state).land_on(args[0]);
    }
    if (this.state.type === StateType.Running && event === Event.Slide) {
      this.state = (<RunningRedHatBoyState>this.state).slide();
    }
    if (this.state.type === StateType.Running && event === Event.KnockOut) {
      this.state = (<RunningRedHatBoyState>this.state).knock_out();
    }
    if (this.state.type === StateType.Running && event === Event.Jump) {
      this.state = (<RunningRedHatBoyState>this.state).jump();
    }
    if (this.state.type === StateType.Sliding && event === Event.Land) {
      this.state = (<SlidingRedHatBoyState>this.state).land_on(args[0]);
    }
    if (this.state.type === StateType.Sliding && event === Event.KnockOut) {
      this.state = (<SlidingRedHatBoyState>this.state).knock_out();
    }
    if (this.state.type === StateType.Jumping && event === Event.Land) {
      this.state = (<JumpingRedHatBoyState>this.state).land_on(args[0]);
    }
    if (this.state.type === StateType.Jumping && event === Event.KnockOut) {
      this.state = (<JumpingRedHatBoyState>this.state).knock_out();
    }
    if (event === Event.Update) {
      this.state = this.state.update();
    }
    return this;
  }

  frame_name(): string {
    return this.state.frame_name();
  }

  context(): RedHatBoyContext {
    return this.state.context;
  }

  update(): RedHatBoyStateMachine {
    return this.transition(Event.Update);
  }

  knocked_out(): boolean {
    return this.state.type === StateType.KnockedOut;
  }
}

interface IRedHatBoy {
  state_machine: RedHatBoyStateMachine;
  sprite_sheet: ISheet;
  image: HTMLImageElement;
}

export class RedHatBoy implements IRedHatBoy {
  state_machine: RedHatBoyStateMachine;
  sprite_sheet: ISheet;
  image: HTMLImageElement;

  constructor(sheet: ISheet, image: HTMLImageElement, audio: Audio, jump_sound: Sound) {
    this.sprite_sheet = sheet;
    this.image = image;
    this.state_machine = new RedHatBoyStateMachine(new IdleRedHatBoyState(audio, jump_sound));
  }

  frame_name(): string {
    return `${this.state_machine.frame_name()} (${Math.floor(this.state_machine.context().frame / 3 + 1)}).png`;
  }

  current_sprite(): ICell {
    return this.sprite_sheet.frames[this.frame_name()];
  }

  draw(renderer: Renderer) {
    let sprite = this.current_sprite();
    renderer.draw_image(
      this.image,
      Rect.new_from_x_y(sprite.frame.x, sprite.frame.y, sprite.frame.w, sprite.frame.h),
      this.destination_box()
    );
  }

  destination_box(): Rect {
    let sprite = this.current_sprite();
    return new Rect(
      this.state_machine.context().position.x + sprite.spriteSourceSize.x,
      this.state_machine.context().position.y + sprite.spriteSourceSize.y,
      sprite.frame.w,
      sprite.frame.h
    );
  }

  bounding_box(): Rect {
    const X_OFFSET: number = 18;
    const Y_OFFSET: number = 14;
    const WIDTH_OFFSET: number = 28;
    let bounding_box = this.destination_box();
    bounding_box.set_x(bounding_box.x() + X_OFFSET);
    bounding_box.width -= WIDTH_OFFSET;
    bounding_box.set_y(bounding_box.y() + Y_OFFSET);
    bounding_box.height -= Y_OFFSET;
    return bounding_box;
  }

  update(): void {
    this.state_machine = this.state_machine.update();
  }

  run_right(): void {
    this.state_machine = this.state_machine.transition(Event.Run);
  }

  slide(): void {
    this.state_machine = this.state_machine.transition(Event.Slide);
  }

  jump(): void {
    this.state_machine = this.state_machine.transition(Event.Jump);
  }

  land_on(position: number) {
    this.state_machine = this.state_machine.transition(Event.Land, position);
  }

  knock_out() {
    this.state_machine = this.state_machine.transition(Event.KnockOut);
  }

  pos_y(): number {
    return this.state_machine.context().position.y;
  }

  velocity_y(): number {
    return this.state_machine.context().velocity.y;
  }

  walking_speed(): number {
    return this.state_machine.context().velocity.x;
  }

  knocked_out(): boolean {
    return this.state_machine.knocked_out();
  }

  static reset(boy: RedHatBoy): RedHatBoy {
    return new RedHatBoy(
      boy.sprite_sheet,
      boy.image,
      boy.state_machine.context().audio,
      boy.state_machine.context().jump_sound
    );
  }
}
